src/btree/bt_ret.c

```diff
@@ -352,20 +307,20 @@ __wt_key_return(WT_CURSOR_BTREE *cbt)
 
 /*
  * __wt_value_return --
- *     Change the cursor to reference an internal return value.
+ *     Change the cursor to reference an update view return value.
  */
 int
-__wt_value_return(WT_CURSOR_BTREE *cbt, WT_UPDATE *upd)
+__wt_value_return(WT_CURSOR_BTREE *cbt, WT_UPDATE_VIEW *upd_view)
 {
     WT_CURSOR *cursor;
 
     cursor = &cbt->iface;
 
     F_CLR(cursor, WT_CURSTD_VALUE_EXT);
-    if (upd == NULL)
+    if (upd_view->type == WT_UPDATE_INVALID)
```

Comment:
Why we need to return anything invalid?

Reply:
I think __value_return is redundant and we can remove that function. Also may be it's worth moving everything in __wt_value_return_upd to __wt_value_return so we can reduce one call.

Reply:
Regarding the "invalid" question, some upper layers of the cursor iteration logic need to differentiate between reading a tombstone or not finding a visible update in order to keep track of some page deletion information.

Reply:
cda8624

Reply:
So invalid means not seeing a visible value I guess. Before durable history, search returns on page value if nothing is visible. I think this is some legacy from before durable history, which is no longer correct. If fixed-length column store interprets it that way, I think the behaviour of fixed-length column may be wrong here. For the page deletion case, I don't understand why we need to return the on page value? Should we just set the cursor to empty value?

Reply:
I remember now fixed length column store is completely broken in durable history because it doesn't have time pairs in the cell. So probably ignore that for now.